### PR TITLE
Create nn_modules_sources structure for mapping attr names to sources

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95831
* #95829
* __->__ #95826
* #95825

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -465,6 +476,23 @@ def wrap_name(module_key):
         for i in itertools.count():
             if name not in self.nn_modules:
                 self.nn_modules[name] = target
+                assert self.nn_modules_sources is not None
+                self.nn_modules_sources[name] = source
+                if isinstance(target, torch.nn.Module):
+                    # annoying, but there are cases when we do not have parameters
+                    # see test_nn_moduledict_contains
+                    if hasattr(target, "_parameters"):
+                        for n, p in target.named_parameters():
```
Comment (wconstab) [inline torch/_dynamo/output_graph.py:485]: so i buy that _parameters may not exist, but does .named_parameters() do something other than iterate an empty set in that case?

**Review by wconstab — APPROVED** [review]

File: torch/_dynamo/output_graph.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:485]: It raises!

**Review by jansel — COMMENTED** [review]
Test failures look real.

**Comment by voznesenskym** [comment]
> Test failures look real.

They are, need to see why.

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -219,6 +223,11 @@ def __init__(
         # should use original graphargs.
         self.orig_graphargs: List[GraphArg] = self.graphargs
         self.nn_modules: Optional[Dict[str, torch.nn.Module]] = dict()
+
+        # A map of normalized names (see normalize_attr_name to their sources)
+        # There is a 1:1 relationship between the module attributes written to the root module
+        # and the names in this map.
+        self.nn_modules_sources: Optional[Dict[str, Source]] = dict()
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:230]: You can't do this. You are requiring a 1:1 relationship, but there is nothing in `normalize_attr_name` that enforces a 1:1 relationship. For example, let's suppose you have an NN module with parameters `foo_bar` and a submodule `foo` which has a parameter `bar`. `foo_bar` and `foo.bar` normalize to the same string, and will conflict in the map here. That's bad!

**Review by ezyang — CHANGES_REQUESTED** [review]
I think my first question for working out an alternative, is why are we not keying off of `source.name()` directly. That key IS guaranteed to be unique and we already use it for hashing guards, e.g.

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:230]: This does use source name for the nested properties, right?

1) Top level register_attr_or_module, uses top level name
2) Iterate over params, make a source `new_source = ParamBufferSource(source, n)`
3) Invoke register_attr_or_module with the source name from 2

```
new_source = ParamBufferSource(source, n)
new_name = new_source.name()
self.register_attr_or_module(p, new_name, source=new_source)
```

I think the problem, as you pointed out, is that normalize_attr_name breaks this uniqueness that sources give us. Let me see if we can maybe find a better way...

We could also probe self.nn_module_sources for if the key exists, and in that case, increment the name to have _0 on it...

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/95825
